### DISCUSSION TITLE
noob openssl in Node.js -- is there an architecture doc somewhere?

### DISCUSSION THREAD

**Discussion by devzzzero** [nodejs/help#4979] [Q&A]
Hi, sorry for the noise, but I am trying to figure out how openssl is integrated into Node.js
Looking at the source code, I see crypto_tls.cc and friends are obvious starters, but I am getting lost in the forest. Specifically, I am getting lost in the NodeJS specific wrappers (e.g. NodeBIO). The NodeJS abstraction layers are thus far a bit of a stumbling block in my comprehension efforts.

Is there a kind soul somewhere that can point me to a birds eye view of how/what parts of openssl work in Node.js?

e.g. here are some of my (admittedly random) questions.

1. What is the significance of using two BIO's? Is it required? (SSL_set_bio man page doesn't seem to state this is the case)
2.  Even though libuv is being used, I don't see any places in the nodejs source code where openssl's non blocking flags are used.
3.  Err.. how is nonblocking thing with libuv actually working? If libuv's event loop is the driver in NodeJS, I'd expect to see libuv callbacks such as ..

void TcpHolder::OnDataReceivedCb(uv_stream_t* tcp, ssize_t nread,
                                 const uv_buf_t* buf) {
  TcpHolder* holder = From(tcp);
  holder->ReclaimUvBuf(buf, nread);
  if (nread < 0 || nread == UV_EOF) {
    holder->handler_->OnEof();
  } else {
    holder->handler_->OnData(&holder->buffer);
  }
}

So holder->handler->OnData .. .where is this redirection set up? And how is node.js working at all with libuv and openssl?
Is there a picture somewhere which describes the overall framework pattern in NodeJS (and how openssl plugs in with libuv's nonblocking io (apparently) without setting SSL_MODE_ASYNC flag anywhere.... ??

Confusion abound. 
Please help!

Thank you
-Jason

**Comment by Trott** [comment]
It doesn't address your questions directly, but if you haven't already read through [the doc on updating OpenSSL in the source](https://github.com/nodejs/node/blob/2849283c4cebbfbf523cc24303941dc36df9332f/doc/contributing/maintaining-openssl.md), it might provide some hints.  And if it doesn't help, it's at least short so shouldn't be a big waste of your time.

**Comment by Trott** [comment]
@nodejs/crypto
